common/src/ipc/transmit.rs

```diff
@@ -288,6 +305,131 @@ impl<T> IpcSocket<T> {
                 .context(IpcErrorKind::MalformedMsg)?;
             Some(Mmap::from_fd(file, len))
         };
-        Ok(RawMsg { code, shm })
+        Ok(IpcMessage { code, shm })
+    }
+}
+
+impl From<Request<'_>> for IpcMessage {
+    fn from(value: Request) -> Self {
+        match value {
+            Request::Ping => Self {
+                code: Code::ReqPing,
+                shm: None,
+            },
+            Request::Query => Self {
+                code: Code::ReqQuery,
+                shm: None,
+            },
+            Request::Kill => Self {
+                code: Code::ReqKill,
+                shm: None,
+            },
+            Request::Clear(clear) => {
+                let mut mmap = Mmap::create(clear.size());
+                clear.serialize(&mut Cursor::new(mmap.slice_mut()));
+                Self {
+                    code: Code::ReqClear,
+                    shm: Some(mmap),
+                }
+            }
+            Request::Img(image) => {
+                let mut mmap = Mmap::create(image.size());
+                image.serialize(&mut Cursor::new(mmap.slice_mut()));
+                Self {
+                    code: Code::ReqImg,
+                    shm: Some(mmap),
+                }
+            }
+        }
+    }
+}
+
+impl<'a> From<&'a IpcMessage> for Request<'a> {
+    fn from(value: &'a IpcMessage) -> Self {
+        match value.code {
+            Code::ReqPing => Self::Ping,
+            Code::ReqQuery => Self::Query,
+            Code::ReqKill => Self::Kill,
+            Code::ReqClear => {
+                let mmap = value.shm.as_ref().expect("clear request must contain data");
```

Comment:
Shouldn't `IPCMessage` be an enum to encode these invariants (of `Clear` and `Img` requiring `shm` to be `Some`?

Reply:
It should, I think I left a todo comment there, just was too lazy to implement it (yet).